plugins: Plugin ideas added

Piankero opened the pull request:
I want to propose 3 new plugin ideas. For the recursion plugin I will need some more advise though from more experienced programmers.

markus2330 commented on src/plugins/port/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Alternative 2
+
+Incorporate the port plugin in the network plugin (with the same behaviour). Those plugins basically do different
```
This makes a lot of sense. More details please (e.g. extend README.md of network plugin).

Btw. actually ping a server would also be interesting but it is a nice-to-have.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+## Introduction
+
+I took [Relax NG](http://relaxng.org/spec-20011203.html#define-ref)'s `ref` and `define` part and tried to "copy" 
+their approach for elektra's specification.
```
I like this idea! Iirc Relax NG is also mathematically sound, could you check if you find a paper with a proof?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+
+I took [Relax NG](http://relaxng.org/spec-20011203.html#define-ref)'s `ref` and `define` part and tried to "copy" 
+their approach for elektra's specification.
+Until now I haven't really gotten a sensible idea for that though. Maybe we could talk on the next meeting about this.
```
I am afraid this is mainly up to you to find out how this works.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+their approach for elektra's specification.
+Until now I haven't really gotten a sensible idea for that though. Maybe we could talk on the next meeting about this.
+
+Also I have no idea what your `struct` plugin actually does. There is no example, just some code snippet from 
```
Thank you for the feedback. The struct plugin was written at a time when no specification existed, so everything was solved via contracts.

The fstab plugin might help you to understand. Important to know is that plugins can read each other's configs/needs.

The idea is that you either extend the struct plugin or you write a new plugin that replaces the struct plugin.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+the contract which doesn't help understanding the plugin.
+
+## Problems
+1) We want to check if all recursions are non-cyclical
```
Details please. Which recursion do you mean? Cycles of links (override/fallback)?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+3) Forcing users to define non-recursive settings first on which they can build up 
+recursive settings is inconvenient and could lead to orphan settings.
+4) We can only check right after calling `kdb set` and not a whole config. 
+This means that user can have "unfinished" configurations while setting new values.
```
I do not understand this limitation. Within "kdb set" you can check the whole configuration and reject everything that is unfinished.

Piankero replied:
That's correct, but this would force the user to have a specified ordering when creating the configuration. Either "bottom-up" (you have to create leaf settings before recursive ones) or "top-down" (you have to create top level settings first before adding lower level ones).

Klemens said that some plugins which take a config do not have a specific ordering once you load them.

markus2330 replied:
Some storage plugins do not remember the order of entries *within the configuration files*. But the order of the configuration structure should have nothing to do with that.

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -0,0 +1,64 @@
+```
+kdb mount config.dump /recursive dump recursion
+kdb setmeta user/recursive check/recursion/vertex "menu, submenu"
+kdb setmeta user/recursive check/recursion/leaf "param, menupoint"
````
Please use Elektra arrays, for example:

```
kdb setmeta user/recursive check/recursion/vertex/#0 menu
kdb setmeta user/recursive check/recursion/vertex/#1 submenu
kdb setmeta user/recursive check/recursion/leaf/#0 param
kdb setmeta user/recursive check/recursion/leaf/#1 menupoint
```

See #1010

A `kdb set-meta-array` would be nice to have.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+
+`check/recursion/vertex` basically means that keys containing `menu` or `submenu` after `/recursive`
+will have further configurations below them. So users would build up structures like this:
+`/recursive/<vertex>/#[0-9]+/<vertex>/#[0-9]+/.....`
```
Can you give a concrete example? Why is `#[0-9]+` allowed?

Piankero replied:
I will change the example in the next commit

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+will have further configurations below them. So users would build up structures like this:
+`/recursive/<vertex>/#[0-9]+/<vertex>/#[0-9]+/.....`
+
+`check/recursion/leaf` means that there is no subvertex allowed anymore. So this is not allowed:
```
What does "param, menupoint" mean?

Piankero replied:
I will change the example in the next commit

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -0,0 +1,64 @@
+
+A command will fail if users want to set settings for non-existent vertices, eg. `kdb set /recursive/menu/#5/....`
+
+The validation solely happens on the structure, not on the values of the keys.
```
Yes, it makes sense to separate this.

Piankero commented on src/plugins/path/README.md:
```diff
@@ -24,9 +25,45 @@ is allowed to occur for both device and mountpoint. When checking for
 a `/`, because remote file systems and some special names are valid, too.
 
+If `check/permission = [permission], [user]` is also present it will check for the correct permissions
+of the file/directory. `check/permission = rw, tomcat` for example will check if the user `tomcat` has read and 
```
What about `check/permission = rwx `and
`check/permission/user = tomcat` in order to split up the permission and ownership?

When the user types in "current" in the user part, I would simply replace it with the user who caused the "kdb set". Does this make sense for you?

markus2330 replied:
> What about check/permission = rwxand
check/permission/user = tomcat in order to split up the permission and ownership?

The ownership has actually little to do with permission.  So maybe `check/owner/user` and `check/owner/group`?

> When the user types in "current" in the user part, I would simply replace it with the user who caused the "kdb set". Does this make sense for you?

This would create a problem if there is a user called "current". What about an empty string? I think this is not allowed as actual user name.

Piankero replied:
My question here is, if the owner is actually that much needed.
I just want to check if a certain user can read/write/etc. on a certain file.

If for example user "java" can write to /var/log/abc.log. Why is the owner of the file of any relevance?

Piankero commented on src/plugins/path/README.md:
```diff
@@ -24,9 +25,45 @@ is allowed to occur for both device and mountpoint. When checking for
+ Permissions available:
+ - `r`: **R**ead
+ - `w`: **W**rite
+ - `x`: e**X**ecute
```
Alright, I will add it in the next commit!

markus2330 replied:
If you agree and have/will fix it, a "thumbs up" :+1: or similar is enough.

Piankero replied:
The sticky bit does not change the behavior of the program at all so I think it is better to ignore it.

I am thinking the same about groups. In the end its a a user who executes or uses the file in the setting. So probably its the best to keep it that simple and only provide rwx.

I am unclear now if you really want to split up permission and user as for me it is more intuitive to have them both in one metadata, eg. check/permission "rwx, java" (or maybe without the colon). If check/permission "rwx, " is given, the current user is taken?

markus2330 replied:
In the implementation it matters: if the file does not exist, it is the sticky bit of the parent directory which may decide if the user can create the file.

Btw. did you check if there an "access (2)" which allows us to say for which user it should be checked? Then the implementation would be trivial.

Maybe there should also be a flag if the file already needs to be present. For log files there are many differences in this area. Actually, if the software does not allow the creation of files, we could simply create it.

> I am unclear now if you really want to split up permission

Yes, the split is a must. Otherwise we get problems if the user names contains ",", spaces or similar.

markus2330 replied:
>  Unfortunately I do not understand what you mean by this confused

I mean to use the syscall access (2) http://man7.org/linux/man-pages/man2/access.2.html

> logs there are rollovers and this could simply lead to a delayed configuration error which isn't wanted.

Yes, this is certainly not wanted. I think we need to gather more data here for what applications do here.

One more advanced idea is that Elektra already opens a file and then passes /proc/self/fd/<nr> instead of the real file name. Then it is impossible that the file disappears (but it needs additionally resources, namely an open file descriptor). 

> What about

Simply propose it in the PR. We can take a look tomorrow.

Piankero commented on src/plugins/path/README.md:
```diff
@@ -24,9 +25,45 @@ is allowed to occur for both device and mountpoint. When checking for
+ - `r`: **R**ead
+ - `w`: **W**rite
+ - `x`: e**X**ecute
+
```
I will leave this open for now and see in my thesis if it is needed that often. But the idea is good!

markus2330 replied:
Yes, making everything complete is not a good idea anyway. It easily happens that stuff is added nobody ever needs. I just mentioned it so that you are aware that such a check might be relevant for some software.

Piankero commented on src/plugins/path/README.md:
```diff
@@ -24,9 +25,45 @@ is allowed to occur for both device and mountpoint. When checking for
+#Generate a file which is only accessable for root
+touch /var/log/application-file-restricted.log
+sudo chmod 700 /var/log/application-file-restricted.log
+sudo chown root:root /var/log/application-file-restricted.log
```
This doesnt trigger an error. It just generates a file with restrictive permissions. Or am I misunderstanding something?

markus2330 replied:
I maybe misunderstood this because of the earlier misunderstanding.

Piankero commented on src/plugins/port/README.md:
```diff
@@ -0,0 +1,27 @@
+
+If `check/port` is specified on a given key, the plugin will validate if the port is a 
+correct number between 1 and 65535.
+Furthermore it will check if the port is open to use.
```
Good Idea, I will add a section with possible enhancements

Piankero commented on src/plugins/port/README.md:
```diff
@@ -0,0 +1,27 @@
+correct number between 1 and 65535.
+Furthermore it will check if the port is open to use.
+
+If `check/simpleport` is specified, only the correct port range will be checked. The port may be in use which is 
```
Indeed, but I didn't come up with a better idea after pondering about this to be honest :P . I will change the behavior as follows:

`check/port/open` Range + available
`check/port/range` Range

I think this is better

markus2330 replied:
What about:

`check/port` for the range check (and/or transformation from /etc/services)
`check/port/listen` to check if the port can be listened to (usually for servers)
`check/port/ping` to check if the port can be pinged/reached  (usually for clients)

Piankero replied:
`check/port/ping` sounds promising but I also think that we first should gather data about the usage. Imo it is often the case the client and server configurations are set before the server is actually started (for example I do it for LCDproc like this). This would prohibit the user from setting a potentially correct setting.

I will add it to a potential future work section

markus2330 replied:
It is also possible and useful to emit warnings.

> I will add it to a potential future work section

If you think it does not work well, it would be rather in the "Discarded Ideas" section.

Piankero commented on src/plugins/port/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Purpose
+
+One source of misconfiguration can come from invalid port numbers or from taking used ports such as 22 for ssh.
```
I just wanted to provide any example. I will change the sentence so it is clear that any port can be used.

markus2330 commented on src/error/specification:
```diff
@@ -1209,3 +1209,9 @@ severity:error
 macro:PLUGIN_TYPECHECKER_SET
+
+number:194
+description:The given path does not meet the correct permissions
```
We should try to reuse/reduce the errors and make a nice classification system. Hopefully you can contribute to improve the error system.

markus2330 commented on src/plugins/network/README.md:
```diff
@@ -37,3 +39,15 @@ it to implement this plugin.
+## Future Work
+
+`check/port/ping` to check if the port can be pinged/reached (usually for clients).
+This potentially could prohibit users from setting correct settings as the server application
```
Not if it is a warning.

The problem is rather that a timeout is needed.

markus2330 commented on src/plugins/path/README.md:
```diff
@@ -26,7 +27,7 @@ relative files, it is not enough to look at the first character if it is
 
 If `check/permission = [permission], [user]` is also present it will check for the correct permissions
-of the file/directory. `check/permission = rw, tomcat` for example will check if the user `tomcat` has read and 
+of the file/directory. `check/permission = rw` and `check/permission/user = tomcat` for example will check if the user `tomcat` has read and 
```
There is still no explanation of check/permission, check/permission/user. It is good to have an example but there should also be an explanation without example.

markus2330 commented on src/plugins/path/README.md:
```diff
@@ -48,10 +49,10 @@ sudo kdb mount some_file.ecf /tests/path some_file dump
 
 #This checks if the file actually exists
-kdb setmeta user/tests/path/value check/path
+sudo kdb setmeta user/tests/path/value check/path
```
"user" needs to be removed, otherwise the change does not make sense.

markus2330 commented on src/plugins/path/README.md:
```diff
@@ -60,10 +61,13 @@ sudo chown root:root /var/log/application-file-restricted.log
-# RET:X
-# ERROR:XXX
+# ERROR:194
+# Reason: User tomcat has no permission to read the given file.
```
Reason should state what was expected and what is actual.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -23,25 +23,37 @@ Users have to define recursive setting specifications in the following way:
 kdb mount config.dump /recursive dump recursion
-kdb setmeta user/recursive check/recursion/vertex "menu, submenu"
-kdb setmeta user/recursive check/recursion/leaf "param, menupoint"
+kdb setmeta user/recursive check/recursion/vertex "A, B"
```
Still without array syntax.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -23,25 +23,37 @@ Users have to define recursive setting specifications in the following way:
 
-`check/recursion/vertex` basically means that keys containing `menu` or `submenu` after `/recursive`
-will have further configurations below them. So users would build up structures like this:
+`check/recursion/vertex` basically means that keys containing `A` or `B` after `/recursive`
```
The word "basically" is useless. Say it like it is. And again: with+without example.

Piankero commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,12 @@ type= enum
+status= idea
+usedby/plugin= path
+type= string
+description= A check to see if a certain user has correct permissions on a file
```
how to specify any combination or rwx?

markus2330 replied:
There are many ways, I am sure you will find one :smile:

Piankero replied:
well, of course I could find one but I have no idea what this file does or how it is used or if it is used. I mean I could provide a regex under "type" or enumerate all possibilities or give a textual description. But I have no idea if I break anything or if I write something illegal then. Is there any documentation for this?

markus2330 replied:
99440157f06eb600b9a90f873dc2e1595bcfe1b0

Piankero commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
 Arbitrary additional data can be saved too such as:
 `kdb set /recursive/define/#2/permission admin`
 
+## A more real example
```
@markus2330 I hope that this example is much more illustrative for you. Please take a look at it and tell me anything which is unclear. It really took me quite some time to get this right

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
 
+## A more real example
+
+In order to better show how this plugin can be used, a more sophisticated example will be given.
```
Its better if the examples show different aspects, not one "less real" and another "more real".

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+## A more real example
+
+In order to better show how this plugin can be used, a more sophisticated example will be given.
+Imagine that you want to create your own menu with arbitrary submenus possible. The following example
```
Menu like in a restaurant?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+The following picture illustrates the menu tree:
+![menu-example](example.png?raw=true "Example")
+
+So how can we map this into an elektra configuration?
```
configuration or specification?

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+```
+kdb mount config.dump /editor dump recursion
+sudo kdb setmeta /editor check/recursion/vertex "menu"
+sudo kdb setmeta /editor check/recursion/ref "menuref"
````
Specification still not explained, what are these two lines for?

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+sudo kdb setmeta /editor check/recursion/vertex "menu"
+sudo kdb setmeta /editor check/recursion/ref "menuref"
+```
+Each key definition in the form of `/editor/menu` will now be treated as recursive entry. You can define arbitrary
````
/editor/menu is not a form but rather an example?

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+sudo kdb setmeta /editor check/recursion/ref "menuref"
+```
+Each key definition in the form of `/editor/menu` will now be treated as recursive entry. You can define arbitrary
+many menues like this: `/editor/menu/#[d]` where `[d]` stands for a positive number.
````
It should be in array form, link to array definition.

Piankero replied:
I cannot find anything in this repo where the "array definition" is explained. Can you give me the link please?

sanssecours replied:
The [array tutorial](https://www.libelektra.org/tutorials/arrays) should provide some basic information about the array syntax.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+many menues like this: `/editor/menu/#[d]` where `[d]` stands for a positive number.
+
+Menues can be linked via `menuref` like this: `/editor/menu/#[d]/menuref/#[d]` and is again in the array notation.
+Note that `menu` and `menuref` are our defined keywords in the setmeta command from above.
```
That should be said earlier. First explain specification or configuration, do not mix up.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+Menues can be linked via `menuref` like this: `/editor/menu/#[d]/menuref/#[d]` and is again in the array notation.
+Note that `menu` and `menuref` are our defined keywords in the setmeta command from above.
+
+Next lets create all distinct menu entries such as `File`, `Print` etc. Note that the Settings
```
let us?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+Note that `menu` and `menuref` are our defined keywords in the setmeta command from above.
+
+Next lets create all distinct menu entries such as `File`, `Print` etc. Note that the Settings
+ menu appears twice but we will only need to declare it once. The order does not matter:
```
order of array entries or order of `kdb set` commands?

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+kdb set /editor/menu/#1/menuref/#2 /editor/menu/#7
+```
+
+Note that `Settings` is already linked. We do not need to do it for the `Tools` menu again.
````
What happens if we link it again?

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+```
+#Link "Macro Basics" to the parent menu "Macros" and create an endless loop
+kdb set /editor/menu/#9/menuref/#0 /editor/menu/#7
+ERR: 198
````
Please use shell recorder syntax

Piankero replied:
Do you have a link to a tutorial for that?

sanssecours replied:
The following documents should be helpful:

- [ReadMe of the Markdown Shell Recorder (MSR)](https://master.libelektra.org/tests/shell/shell_recorder/tutorial_wrapper/README.md)
- [the MSR syntax check](https://master.libelektra.org/tests/shell/shell_recorder/tutorial_wrapper/SyntaxCheck.md)

. Many of the plugins also contain some basic MSR tests. For example, the ReadMe of [YAML CPP](https://www.libelektra.org/plugins/yamlcpp) contains a few example tests.

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+#Link to something which does not exist
+kdb set /editor/menu/#9/menuref/#0 /editor/menu/#10
+ERR: 199
+Reason: Could not find vertex with #10. Maybe it does not exist yet?
```
Why isn't the term also used in the tutorial?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+Their approach is not array based but name based. 
+Their reference keyword (`check/recursion/ref`) to other settings is `Entry`.
+
+They do not use an array based approach though but give names instead.
```
Can we add a name-based approach or does it confict with the array-based approach?

Piankero replied:
It doesn't conflicht imo, as discussed in your comment below it needs to know if it is a vertex or some arbitrary setting unrelated to the recursive structure. With the array based approach this is clear.

markus2330 replied:
Imho, the array does not make a difference. Without some external knowledge you never know if a string is a reference or a value interpreted otherwise.

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+kdb set /editor/AutocorrectOptions/DisplayName "Autocorrect Options"
+``` 
+
+This would at least guarantee reusability of settings.
````
Why "at least"? What is missing?

markus2330 commented on src/plugins/recursion/README.md:
```diff
@@ -81,10 +81,186 @@ kdb set /recursive/define/#2/ref/#1 /recursive/define/#5
+This would at least guarantee reusability of settings.
+
+The plugin could theoretically support an arbitrary name instead of an array. But how does the 
+specification know what is actually a menu entry and what is just some arbitrary other setting.
```
Metadata?

Piankero replied:
How could this look like?

Piankero commented:
I updated the documentation.
On sunday I will
1) Add src/error/specification for port & recursion
2) write the correct shell recorder syntax

Concerning the "named" recursion it would be interesting to have your feedback.
Also on the `Problem` part in the recursion plugin readme which I updated.

Piankero commented on src/plugins/recursion/README.md:
````diff
@@ -10,18 +10,37 @@ Also I have no idea what your `struct` plugin actually does. There is no example
+[b]
+```
+
+which could cause an error since `a` cannot be loaded as `b` is yet non existent.
````
Can you explain why? How exactly are keys loaded into the kdb?

I assume that it is loaded in the order of the file, i.e. it would be the same if I call kdb set ... in the same order as the config file

markus2330 replied:
The plugins receive a whole keyset and can do whatever they want to do with it. So in the implementation you can lookup if something is present, it does not matter where in the keyset it is.

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+status= idea
+usedby/plugin= path
+type= string
+	empty
```
empty or string?

Piankero replied:
either or. If empty then the current user is taken, if a string is given it is interpreted as the desired user. It is also written in the description

markus2330 replied:
string automatically includes empty. Elektra does not have sum types, so specifying "string empty" does not make sense. (The result would be "empty")

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+usedby/plugin= path
+type= string
+	empty
+description= Used be check/permission to see if the given user has the correct permissions. If empty
```
empty or non-existent?

Piankero replied:
Isn't it the same question as above?

markus2330 replied:
Yes, it is the same problem.

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+type= string
+	empty
+description= Used be check/permission to see if the given user has the correct permissions. If empty
+	it takes the current user.
```
Can it be used without check/permission? Maybe assume rwx?

Piankero replied:
Yes but I think it would lead to more errors since for example `x` is rarely used and attached to a file. 
For a default value I would suggest only `r`.

Still I think it's the best to let the user decide which permissions should be possible as it is very different from file to file. Maybe when we gather more data it is clear?

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+status = idea
+usedby/plugin = network
+type = empty
+desciption = If specified, checks if the value of the key is between 0 and 65535.
```
As non-root it is 1025-65535. Without checking this it would make little sense to have a separate check? (check/type could be used).

And the translation from port name -> number should also be done (done by getaddrinfo anyway, see below)

Piankero replied:
What about emitting a warning when trying to set a port below 1025? I think the usability would suffer if users would need to add a separate check just to be assured that a port below 1025 is executed as root.

Also how exactly would you think of check/type? Can you give an example?

markus2330 replied:
You are right, the root check only makes sense for servers, not for clients. So there are 3 ways:

1. If a user simply wants a check from 0-65535, she would use `check/type=unsigned_short`.
2. If a user wants to check if a string can be used as port (number 0-65535 or a name from /etc/services), she would use check/port
3. If a user wants to check if a string can be used as port to listen to (as above + not occupied + root for 0-1024), she would use check/port/listen

Btw. is 0 a valid port number?

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+type = string
+decription = Used to declare recursive structures like trees or graphs. Every key
+			  below the current key will be treated as vertex from which you can refer to by given string.
+example = eg if 'Entry' was specified, one could refer to other vertices by your/key/Entry/#0 <vertex key>
```
Here should be only an example.

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+[check/port]
+status = idea
+usedby/plugin = network
+type = empty
```
It is a number isn't it?

Piankero replied:
check/port does only a check if the range is correct. A number is not required imo. type is only used to declare the type of the metavalue

markus2330 commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+status = idea
+usedby/plugin = network
+type = empty
+desciption = If specified, checks if the given port is open and ready to be listened to.
```
relation to check/port not documented

markus2330 commented on src/error/specification:
```diff
@@ -1209,3 +1209,39 @@ severity:error
+module:path
+
+number:196
+description:The given port is not in a valid range (0-65535)
```
The range can be given in the reason.

markus2330 commented on src/error/specification:
```diff
@@ -1209,3 +1209,39 @@ severity:error
+module:network
+
+number:196
+description:The desired port is already in use.
```
Please be consistent with the already existing errors if you put a . in the end or not.

Piankero replied:
Some descriptions have a dot and some do not. Which one should i take now?

markus2330 replied:
Ok, this is something to unify later.

We can also make that sentences get a .

markus2330 commented on src/error/specification:
```diff
@@ -1209,3 +1209,39 @@ severity:error
+description:Referring to the desired key would lead to an endless loop
+severity:error
+ingroup:plugin
+module:network
```
network or recursive?

Piankero replied:
woops, copy and paste error

markus2330 commented on src/plugins/network/README.md:
```diff
@@ -14,6 +14,8 @@ This plugin is a check plugin that checks if a key contains a valid ip
 address. It uses the `POSIX.1-2001` interface `getaddrinfo()` in order
 to check if an ip address is valid.
 
+The plugin can also check for valid port numbers and if the set port is free to use.
```
The port should actually also passed to getaddrinfo. Better to rewrite the intro.

markus2330 commented on src/plugins/network/README.md:
```diff
@@ -37,3 +39,14 @@ it to implement this plugin.
+
+## Future Work
+
+`check/port/ping` to check if the port can be pinged/reached (usually for clients).
```
maybe connect is better, as ping somehow implies ICMP.

markus2330 commented on src/plugins/recursion/README.md:
````diff
@@ -75,30 +66,33 @@ Users will have to set this structure in the following way:
 ```
 
+It is not needed to add `/recursive` before every key as it is only possible to choose keys
+below this rootkey anyway. Referring to other keys is done by 
````
directly below?

Piankero commented on doc/METADATA.ini:
```diff
@@ -679,6 +679,43 @@ type= enum
+description= A check to see if a certain user has correct permissions on a file. Permissions are
+	either r (read) w (write) x (execute) and any combination of them.
+	 The respective user which must have these permissions is declared in check/permission/user.
+example = "rwx ; rw ; r ; x ; w ; wx"
```
What is a valid example? They are all just some descriptions as i can see. I made it more verbose now

markus2330 replied:
The example needs to be something that actually could be a value of the metavalue.

E.g. "rwx" would be a valid example.

markus2330 commented on doc/METADATA.ini:
```diff
@@ -682,17 +682,24 @@ description= Support of some standards for a date specification.
 status= idea
 usedby/plugin= path
-type= enum
+type= r
```
enum and then the possibilities

markus2330 commented on src/error/specification:
```diff
@@ -1238,10 +1238,10 @@ number:197
 ingroup:plugin
-module:network
\ No newline at end of file
+module:recursion
```
Its better to have a new-line in the end of files.

markus2330 commented on src/plugins/network/README.md:
```diff
@@ -14,7 +14,9 @@ This plugin is a check plugin that checks if a key contains a valid ip
 
-The plugin can also check for valid port numbers and if the set port is free to use.
+Furthermore `getaddrinfo()` is used in `check/port` to resolve a port by its service name
+which is defined under `/etc/services`.
```
and it also transforms to the number?

Piankero commented:
Hopefully now everything is accordingly.

Could you take a look at the new recursion plugin? It just contains a single check metadata now instead of 2.

kodebach commented:
@markus2330 I updated the recursion plugin README with results from a talk between me and @Piankero 

I would also suggest renaming the recursion plugin to `reference`, because one major part of the work the proposed plugin would do is checking, whether references are valid. This functionality could also be useful elsewhere without the recursive part.

I propose:
* changing `check/recursion` to `check/reference`
* changing `check/recursion/override` to `check/reference/override`
* changing `check/recursion/restrict` to `check/reference/restrict`
* adding the boolean `check/reference/recursive` to enable/disable the check for loops in the reference graph

An example of such reference checking without recursion would be the references to driver configurations in the lcdproc server configuration.

I would also suggest either closing this PR entirely and opening separate PRs for the proposed plugins, or at least opening a separate PR for the recursion/reference plugin, to keep things a bit more organized.

markus2330 commented:
Yes, please let us open a new PR once we decided on the new name. If the checks are not limited to recursive checks, it makes sense to rename it. How can a user distinguish if a reference check is recursive or only a single check? What if we remove the old "struct" check and call the new plugin also "struct"? Is this plugin able to check some arbitrary recursive structure of configuration?

Btw. if possible we should avoid "override" because this already has a meaning within Elektra (it is a link to be preferred, see doc/METADATA.ini).

kodebach commented:
> How can a user distinguish if a reference check is recursive or only a single check?
```ini
[/tests/keywithref]
check/reference = ref

[/tests/keywithrec]
check/reference = ref
check/reference/recursive = true
```
Here `/tests/keywithref/ref` would just be a single reference, while `/tests/keywithrec/ref` would be checked recursively. 

An alternative would be the following:

```ini
[/tests/keywithref/ref]
check/reference = single

[/tests/keywithrec/ref]
check/reference = recursive
```
Here the type of `check/reference` would be `enum`, with the values `single` and `recursive` being supported. The name used for recursive checking would be the basename of the key with the `check/recursive` metadata. Then overriding the name could be done through a third enum value.

Personally I like the second option better. It makes more sense to me to specify the metadata on the actual reference key.

> What if we remove the old "struct" check and call the new plugin also "struct"? Is this plugin able to check some arbitrary recursive structure of configuration?

The current proposed plugin would only check references. Checking the remaining structure could simply be done with the `spec` plugin:

```ini
[/tests/base/ref]
check/reference = recursive
check/reference/restrict = typeA

[/tests/base/typeA/_/name]
check/type = string
; etc

[/tests/base/typeA/_/size]
check/type = long
; etc
```
The basic idea is to restrict the references to a certain path `X` and then using `X` as a kind of type identifier. The direct sub-keys of `X` would then be 'instances' of `X`, so by using `X/_/Y` one can specify properties of the field `Y` of type`X`.

markus2330 commented:
> Personally I like the second option better. It makes more sense to me to specify the metadata on the actual reference key.

Yes, I also like the second option better.

> The basic idea is to restrict...

Good example. Could you please incorporate it into the README? But please continue using typeA (if I understood correctly it is changed to X later in the explanation).

Piankero commented:
@markus2330 
There is a problem with the extension of the `path` plugin and the assumption if the metakey`check/permission/user` is not set, the current user is taken.

The problem lies in the `euidaccess` function and the need to change the egid.

Is is a bit difficult to explain:\
*) `euidaccess` checks for the access with the egid and euid.
*) unfortunately there can only be one egid active at a time when calling `euidaccess()`.
*) So if a user belongs to multiple groups and the filegroup is one of them, I have to actively switch to this group with `setegid()` before calling `euidaccess()`
*) To switch via `setegid()` I have to be running as root even though I am a member of the group I want to "switch" to.
*) When running as root though, the information is lost for which user you really want to check the permission for.

So it is absolutely required that check/permission/user and check/permission/types are both present.
I can only support to assume if check/permission/user is empty the root user is assumed. But checking for file access as root user is a bit fruitless.

I hope I could explain the problem so that it is understandable. Do you have a better solution or do we simply require the coexistence of both metadatas?

markus2330 commented:
What is "check/permission/types"?

> But checking for file access as root user is a bit fruitless.

I am not sure if it is fruitless. There are now many kernel extensions (and also file system features) to disallow root to do something. And I would prefer the "current user" and not root as default. And in the "current user" scenario no switching of egid is necessary.

If two options are absolutely required so that the plugin can do something useful it is of course possible to demand the presence of both options. Afaik this would be the first plugin, though.

Piankero commented:
> What is "check/permission/types"?

 It is used for the actual permissions, e.g. "rwx"

> And I would prefer the "current user" and not root as default. And in the "current user" scenario no switching of egid is necessary.

I wish for the same but when programming I could not circumvent the problem described above. But I have to switch egid since access only checks with one egid active. If you wish I could also demonstrate it on our next Elektra meeting.

Piankero commented:
All plugin ideas are now in implementation stage in their respective PR. Closing this PR
